contents/linux/shell.md

```diff
@@ -0,0 +1,180 @@
+Shell is the interface through which we can pass instructions to kernel, and then it could be executed on the hardware. In this chapter, we will look into **Linux command-line based shell(bash)**.
+
+> Linux is the best-known and most-used open source operating system. As an operating system, Linux is software that sits underneath all of the other software on a computer, receiving requests from those programs and relaying these requests to the computer’s hardware.
+
+[**bash**](https://www.gnu.org/software/bash/manual/html_node/) stands for **Bourne Again SHell**, an enhanced version of the original Unix Shell program. It has become the standard shell of various linux distributions. 
+
+## Why use command-line based shell?
+
+A common question many people would ask is why would we learn command-line based shell as it seems to be very hard to use. There are tons of GUI tools out there, they seems to look more user-friendly and powerful. From my perspective, there are four major reasons for starting to use command line.
+
+- **Remote Management**: Use over remote command-line shell makes it much easier to perform tasks remotely, such as when a developer is connecting from home. A remote command-line shell usually requires less bandwidth.
+
+- **Stability**: No matter what linux distribution you are running, the command line shell is always available and usage is almost the same. However, GUI tools might be different in different distributions, make it very hard to learn.
+
+- **Freedom**: You will have complete control in command-line shell. GUIs usually tends to simplifying things, giving user fewer options to make things easier. In command line there is no such simplifications, you will have complete control of what you are doing.
+
+- **Simplicity**: Although you might find command-line hard to use at first, but after using it enough you'll eventually come to the point where the command-line is far easier to use than trying to find options in the drop-down menus and alike of a GUI. 
+
+### Benefit: Program with bash script
+
+One important feature in bash is **bash script**. It enables us to do `task automation`. Suppose you are a system admin, there is a task you need to perform daily without any changes. Instead of typing in the set of commands for the task everyday, you can store these commands in a file and run the script.
+
+Here's one example of bash script, suppose you have to check if some given file exist periodically. The following script will ask for your input and output the results. `filename` is a variable used to store user input, `filetype` and `perm` stores corresponding attributes of the file for output.
+
+```bash
+#!/bin/bash
+ 
+echo -e "Please input a filename, I will check the filename's type and permission.\n\n"
+
+# waiting for user input.
+read -p "Input a filename : "  filename
+
+# check if file exists.
+test -z $filename && echo "You MUST input a filename." && exit 0
```

Comment:
The Google Shell Style Guide [recommends `if [[...]]` for tests](https://google.github.io/styleguide/shell.xml?showone=Test,_%5B_and_%5B%5B#Test,_%5B_and_%5B%5B).